telemetry/telemetry.go

```diff
@@ -211,6 +213,57 @@ func isFlagPassed(fs *flag.FlagSet, name string) bool {
 	return found
 }
 
+func iNotifyCertMonitoring(watcher *fsnotify.Watcher, telemetryCfg *TelemetryConfig, serverControlSignal chan<- int, testReadySignal chan<- int) {
+	defer watcher.Close()
+
+	done := make(chan bool)
+	go func() {
+		if testReadySignal != nil { // for testing only
+			testReadySignal <- 0
+		}
+		for {
+			select {
+			case event := <-watcher.Events:
+				if event.Name != "" {
+					log.V(1).Infof("Inotify watcher has received event: %v", event)
+					if event.Op & fsnotify.Write == fsnotify.Write {
+						log.V(1).Infof("Cert File has been modified: %s", event.Name)
+						serverControlSignal <- 1 // tells gnmi server to stop and restart with new certs
+						done <- true
+						return
+					}
+					if event.Op & fsnotify.Remove == fsnotify.Remove {
+						log.Errorf("Cert file has been deleted: %s", event.Name)
+						serverControlSignal <- 0 // tells gnmi server to stop
```

Comment:
>stop [](http://example.com/codeflow?start=55&length=4)

Is it force stopping or graceful stopping in your design?

Reply:
We may design both, but implement at least the force stopping if that is minimal required.

Reply:
ServerControlSignal channel receives a value when SIGKILL or SIGTERM is received, monitored cert file has been deleted or changed, cert watcher has received a non-filesystem error.

When serverControlSignal channel receives a value, the running gnmi server will always be force stopped via s.Stop(). And depending on the value, gnmi server will either be restarted or not. If cert is rotated, we will restart gnmi server, but in the other cases we will let the goroutine return.

https://pkg.go.dev/google.golang.org/grpc#Server.Stop

"Stop stops the gRPC server. It immediately closes all open connections and listeners. It cancels all active RPCs on the server side and the corresponding pending RPCs on the client side will get notified by connection errors."

Reply:
so it is force stop.